### PR TITLE
[PUBDEV-4495] Fix bug with self node plus LOG refactoring and unification

### PR DESCRIPTION
**Pull request by jakubhava**
This change also introduces Types for LogLevels preventing us to set incorrect ones -> less exceptions to be thrown

### CODE REVIEW

**Comment by mmalohlava** [comment]
@jakubhava this needs little bit of explanation here

@abal5 this is java change, but Java tests were not triggered.

**Comment by jakubhava** [comment]
@mmalohlava from commit description
- Fix bug when getting logs from local node
- Ensure the logging is initialized properly. In previous code one of the methods Log.Info bypassed initialization and tried toi log right away
- Don't throw exception when log file does not exist right away, but first check whether we logs on that level are enabled
- Capture a few missing exceptions and remove extra ones in Log.class
- Refactor and clean up Log.class.
- Put logs into unique directories so they don't overlap

I decided the clean the code a little bit when doing this change.

**Comment by mmalohlava** [comment]
Seems like lot of changes, but where was actual bug?

**Comment by jakubhava** [comment]
There was more bugs 
- Fix bug when getting logs from local node - current node `-1` wasn't handled properly
- Logs where tried to be obtained even for logs with higher log level
- for Log.Info method the initialisation of logging was skipped and therefore it could cause NPE if somebody would use it as a first log method
- in LogHandler, name for log file was obtained without check if the logging was initialised

The rest is just refactorization. I can change this PR just to fix the errors above, though I thought improving the code ( already when touching it ) can't hurt

**Comment by mmalohlava** [comment]
:+1:

**Comment by jakubhava** [comment]
However I'm opened to reverting some stuff, it's true that it's big change. 

Oh and the additional parse check in H2O might be moved to another commit, or we may drop it completely. I tried to pass negative or higher number and the initialisation of H2O failed with a weird error

**Comment by mmalohlava** [comment]
Kuba, it is fine - i was just trying understand scope of change. Refactoring is ok here.

**Comment by jakubhava** [comment]
I split this change into several commit so it's easier to review

File: h2o-core/src/main/java/water/api/LogsHandler.java [hunk 1]
```diff
@@ -114,45 +115,43 @@ else if (  name.equals("trace")
 
 
   private static H2ONode getH2ONode(String nodeIdx){
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/api/LogsHandler.java]: Conceptually, you can split the code into two methods:

```
H2ONode getH2ONode(String nodeRef) {
  if (nodeRef is index) getH2ONodeByIndex(nodRef)
  else if (nodeRef is ip:port) getH2ONodeByIpPort(ip, port)
  else error("Unknown node reference)
}```

File: h2o-core/src/main/java/water/util/Log.java [hunk 2]
```diff
@@ -22,261 +23,234 @@
  *  logging, no guarantees are made about the messages.
  **/
 abstract public class Log {
+  private static final byte UNKNOWN = -1;
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/util/Log.java]: Why we need constant and enum?

File: h2o-core/src/main/java/water/util/Log.java [hunk 3]
```diff
@@ -22,261 +23,234 @@
...
-      INIT_MSGS.clear();
+      initialMsgs.clear();
     }
   }
 
-  /**
-   * @return This is what should be used when doing Download All Logs.
-   */
+  /** Get directory where the logs are stored */
   public static String getLogDir() throws Exception {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/util/Log.java]: Does not make sense to throw untyped exception, in this case we can simply throws RuntimeException and remove `trows` from method signature.

File: h2o-core/src/main/java/water/util/Log.java [hunk 4]
```diff
@@ -22,261 +23,234 @@
...
+    return LEVEL.fromNum(currentLevel);
   }
 
-  private static String getLogFileNameStem() throws Exception {
-    String ip = H2O.SELF_ADDRESS.getHostAddress();
-    int port = H2O.API_PORT;
-    String portString = Integer.toString(port);
-    return "h2o_" + ip + "_" + portString;
+  /** Get full path to a log file of specified log level */
+  public static String getLogFilePath(String level) throws Exception {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/util/Log.java]: The same as above - we can remove `throws` from the method signature.

File: h2o-core/src/main/java/water/util/Log.java [hunk 5]
```diff
@@ -288,7 +262,7 @@ private static void setLog4jProperties(String logDir, java.util.Properties p) th
 
     p.setProperty("log4j.appender.HTTPD",                       "org.apache.log4j.RollingFileAppender");
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/util/Log.java]: Aren't all block same except `Rx` and name of log level and its index? We can do DRY here... ;-)

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@mmalohlava thanks for all the comments! I will work on the fixes tonight and will try to create tests for it

File: h2o-core/src/main/java/water/util/Log.java (hunk 2 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/util/Log.java]: Yes, that was my first approach. However I wasn't sure about the performance. I wanted to replace numeric log levels by enums. Do you think it's fine to do ? I would like that better as well.

Comment (mmalohlava) [inline h2o-core/src/main/java/water/util/Log.java]: yes, it is fine

**Comment by jakubhava** [comment]
Manually tested on Hadoop and locally ( both viewLogs and DownloadLogs), works as expected

**Comment by michalkurka** [comment]
@jakubhava would it be possible to split this into 2 PRs? one bugfix, one refactoring. This is hard to read.

**Comment by jakubhava** [comment]
There's several bug fixes actually, please read the PR comments. I can extract this into separate PR if that helps -> https://github.com/h2oai/h2o-3/pull/1186/commits/95863c0f24dae7bd5ce4af878e00928f0f8acdba, the rest are related to the refactoring. The code was using lots of unchecked exceptions

**Comment by jakubhava** [comment]
@michalkurka btw, I was creating that many commits exactly for this reason - so we can read the one after another to see the progress. However if it's blocking you from doing review I can try to separate it into more separated PRs and base each on top of another. The changes are in separated commits. However the refactoring will be ugly anyway.

**Comment by michalkurka** [comment]
@jakubhava thanks, this is fine, I will read the commits

**Comment by jakubhava** [comment]
But you're right. Next time I'm doing anything similar I'll try completely separate bug fixes and refactoring. Thanks for pointing that out :)

File: h2o-core/src/main/java/water/api/LogsHandler.java (hunk 1 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/api/LogsHandler.java]: Separated in the latest version

File: h2o-core/src/main/java/water/util/Log.java (hunk 3 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/util/Log.java]: good point - done

File: h2o-core/src/main/java/water/util/Log.java (hunk 5 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/util/Log.java]: Good point, refactored in the last version

**Comment by jakubhava** [comment]
@h2o-ops please test!

File: h2o-core/src/main/java/water/util/Log.java [hunk 6]
```diff
@@ -22,277 +27,205 @@
...
   }
 
-  private static void setLog4jProperties(String logDir, java.util.Properties p) throws Exception {
-    LOG_DIR = logDir;
-    String logPathFileName = getLogPathFileNameStem();
+  private static void setPropertiesFor(Properties p, String logger, String pattern, String level, String logPath){
+    p.setProperty("log4j.appender." + logger,                               "org.apache.log4j.RollingFileAppender");
+    p.setProperty("log4j.appender." + logger + ".Threshold",                level);
+    p.setProperty("log4j.appender." + logger + ".File",                     logPath);
+    p.setProperty("log4j.appender." + logger + ".MaxFileSize",              "1MB");
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/Log.java]: It looks like the MaxFileSize was unified to 1MB, is this intended?

File: h2o-core/src/main/java/water/util/Log.java [hunk 7]
```diff
@@ -22,277 +27,205 @@
...
+            "TRACE, R5, R6",
+            "TRACE, R4, R5, R6",
+            "TRACE, R3, R4, R5, R6",
+            "TRACE, R2, R3, R4, R5, R6",
+            "TRACE, R1, R2, R3, R4, R5, R6",
+    }[currentLevel.numLevel];
     p.setProperty("log4j.logger.water.default", appenders);
     p.setProperty("log4j.additivity.water.default",   "false");
 
-    p.setProperty("log4j.appender.R1",                          "org.apache.log4j.RollingFileAppender");
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/Log.java]: I personally think that the previous was clearer than the new code. It was essentially a properties file represented in Java. Now it is harder to figure out what is going on.

**Comment by jakubhava** [comment]
@michalkurka I was thinking a bit and I could make separate this change at the end as it's really not good idea to put refactoring and bug fixes into one PR. We can either survive this PR and I make sure to do the separation in the future or separate this one. What do you think ?

File: h2o-core/src/main/java/water/util/Log.java (hunk 6 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/util/Log.java]: Good catch! This was unwanted

File: h2o-core/src/main/java/water/util/Log.java (hunk 7 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/util/Log.java]: I implemented it based on @mmalohlava suggestion. The code was kind of repetitive so wanted to get rid of that and have one single place where we configure it.

I can however change it back, I have no strong opinion on either original or new approach. The original is easier to read, to second one is less code with single point of configuration.

Comment (jakubhava) [inline h2o-core/src/main/java/water/util/Log.java]: Resolved in the last commit. Good point regarding the consistency with the regular log4j configurationResolved in the last commit. Good point regarding the consistency with the regular log4j configuration

**Comment by michalkurka** [comment]
@jakubhava I think it would be very helpful to split it into two PRs. If you could do that that would be awesome

**Comment by jakubhava** [comment]
@michalkurka oki I'll separate it :)

**Pull request closed without merging**